devito/passes/iet/misc.py

```diff
@@ -74,49 +91,80 @@ def relax_incr_dimensions(iet, **kwargs):
-
-        # Remove any offsets
-        # E.g., `x = x_m + 2 to x_M - 2` --> `x = x_m to x_M`
-        outer = [i._rebuild(limits=(i.dim.root.symbolic_min, i.dim.root.symbolic_max,
-                                    i.step))
-                 for i in outer]
-
-        # Create the ElementalFunction
-        name = sregistry.make_name(prefix="bf")
-        body = compose_nodes(outer)
-        dynamic_parameters = flatten((i.symbolic_bounds, i.step) for i in outer)
-        dynamic_parameters.extend([i.step for i in inner if not is_integer(i.step)])
-        efunc = make_efunc(name, body, dynamic_parameters)
-
-        efuncs.append(efunc)
-
-        # Create the ElementalCalls
-        calls = []
-        for p in product(*ranges):
-            dynamic_args_mapper = {}
-            for i, (m, M, b) in zip(outer, p):
-                dynamic_args_mapper[i.symbolic_min] = m
-                dynamic_args_mapper[i.symbolic_max] = M
-                dynamic_args_mapper[i.step] = b
-                for j in inner:
-                    if j.dim.root is i.dim.root and not is_integer(j.step):
-                        value = j.step if b is i.step else b
-                        dynamic_args_mapper[j.step] = (value,)
-            calls.append(efunc.make_call(dynamic_args_mapper))
-
-        mapper[root] = List(body=calls)
-
-    iet = Transformer(mapper).visit(iet)
-
-    return iet, {'efuncs': efuncs}
+        # Get root symbolic_max out of each outer dimension
+        roots_max = {i.dim.root: i.symbolic_max for i in outer}
+
+        # A dictionary to map maximum of processed parent dimensions
+        # Helps to neatly handle bounds in hierarchical blocking
+        proc_parents_max = {}
+
+        # Process inner iterations
+        for n, i in enumerate(inner):
+            if i.dim.parent in proc_parents_max and i.symbolic_size == i.dim.parent.step:
+                # In hierarchical blocking (BLOCKLEVELS > 1) a parent dimension may
+                # have already been processed as an 'inner' Iteration. Since in
+                # hierarchical blocking block sizes in lower levels always perfectly
+                # divide block sizes of upper levels we can use the parent's Iteration
+                # maximum
+                iter_max = proc_parents_max[i.dim.parent]
+            else:
+                # Candidate 1: symbolic_max of current iteration
+                # Most of the cases pass though this code:
+                # e.g. `i.symbolic_max = x0_blk0 + x0_blk0_size`
+                symbolic_max = i.symbolic_max
+
+                # Candidate 2: maximum of the root iteration
+                # Candidate 2a: Usually the max of parent/root dimension, e.g. `x_M`
+                root_max = roots_max[i.dim.root]
+
+                # Candidate 2b:
+                # the `symbolic_size` of an Iteration may exceed the size of a parent's
+                # block size (e.g. after CIRE passes) e.g.
```

Comment:
I think, though may be wrong, that this comment is written in a misleading way. 

if an Iteration "may exceed the size of a parent's block size" , then why would you ever omit iterations? is it really not to omit iterations, or to avoid going out of bounds?